File: protocol/native-v1alpha1/conformance/canonical.go
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// Standalone canonical JSON adapter; schema/semantic validation is in check.py.
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strconv"
)

func printableASCII(value string) error {
	for _, r := range value {
		if r < 32 || r > 126 {
			return fmt.Errorf("non-printable ASCII")
		}
	}
	return nil
}

// Token parsing rejects duplicate keys, which decoding into a map would hide.
func parse(d *json.Decoder) (interface{}, error) {
	t, err := d.Token()
	if err != nil {
		return nil, err
	}
	switch v := t.(type) {
	case json.Delim:
		if v == '{' {
			m := map[string]interface{}{}
			for d.More() {
				key, e := d.Token()
				if e != nil {
					return nil, e
				}
				s, ok := key.(string)
				if !ok {
					return nil, fmt.Errorf("object key")
				}
				if e := printableASCII(s); e != nil {
					return nil, e
				}
				if _, exists := m[s]; exists {
					return nil, fmt.Errorf("duplicate key")
				}
				item, e := parse(d)
				if e != nil {
					return nil, e
				}
				m[s] = item
			}
			end, e := d.Token()
			if e != nil || end != json.Delim('}') {
				return nil, fmt.Errorf("object end")
			}
			return m, nil
		}
		if v == '[' {
			a := []interface{}{}
			for d.More() {
				item, e := parse(d)
				if e != nil {
					return nil, e
				}
				a = append(a, item)
			}
			end, e := d.Token()
			if e != nil || end != json.Delim(']') {
				return nil, fmt.Errorf("array end")
			}
			return a, nil
		}
		return nil, fmt.Errorf("delimiter")
	case json.Number:
		n, e := strconv.ParseUint(string(v), 10, 64)
		if e != nil || n > 9007199254740991 {
			return nil, fmt.Errorf("noncanonical integer")
		}
		return n, nil
	case string:
		if e := printableASCII(v); e != nil {
			return nil, e
		}
		return v, nil
	default:
		return t, nil
	}
}
func main() {
	if len(os.Args) != 2 {
		panic("usage: canonical document.json")
	}
	data, err := os.ReadFile(os.Args[1])
	if err != nil {
		panic(err)
	}
	d := json.NewDecoder(bytes.NewReader(data))
	d.UseNumber()
	value, err := parse(d)
	if err != nil {
		panic(err)
	}
	if _, err = d.Token(); err != io.EOF {
		panic("trailing input")
	}
	encoder := json.NewEncoder(os.Stdout)
	encoder.SetEscapeHTML(false)
	if err = encoder.Encode(value); err != nil {
		panic(err)
	}
}

File: protocol/native-v1alpha1/conformance/Canonical.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/* Standalone Java 8 canonical adapter, deliberately independent of legacy Gradle.
 * This is a bounded fixture parser, not a production protocol implementation.
 * Schema and semantic validation remain in check.py. */
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public final class Canonical {
  private final String text;
  private int offset;
  private Canonical(String text) { this.text = text; }
  private IllegalArgumentException error() {
    return new IllegalArgumentException("Invalid bounded JSON at " + offset);
  }
  private void space() {
    while (offset < text.length() && " \t\n\r".indexOf(text.charAt(offset)) >= 0) offset++;
  }
  private char take() { if (offset == text.length()) throw error(); return text.charAt(offset++); }
  private boolean eat(char c) {
    space(); if (offset < text.length() && text.charAt(offset) == c) { offset++; return true; }
    return false;
  }
  private String string() {
    if (take() != '"') throw error();
    StringBuilder result = new StringBuilder();
    while (true) {
      char c = take();
      if (c == '"') return result.toString();
      if (c == '\\') {
        c = take();
        if (c == 'u') {
          if (offset + 4 > text.length()) throw error();
          for (int index = offset; index < offset + 4; index++) {
            if ("0123456789abcdefABCDEF".indexOf(text.charAt(index)) < 0) throw error();
          }
          c = (char) Integer.parseInt(text.substring(offset, offset + 4), 16); offset += 4;
        } else if (c != '"' && c != '\\' && c != '/') throw error();
      }
      if (c < 32 || c > 126) throw error();
      result.append(c);
    }
  }
  private Object value() {
    space(); if (offset == text.length()) throw error();
    char c = text.charAt(offset);
    if (c == '"') return string();
    if (eat('{')) {
      Map<String, Object> map = new TreeMap<>();
      if (eat('}')) return map;
      do {
        space(); String key = string(); if (!eat(':') || map.containsKey(key)) throw error();
        map.put(key, value());
      } while (eat(','));
      if (!eat('}')) throw error(); return map;
    }
    if (eat('[')) {
      List<Object> list = new ArrayList<>(); if (eat(']')) return list;
      do { list.add(value()); } while (eat(','));
      if (!eat(']')) throw error(); return list;
    }
    for (String literal : new String[]{"true", "false", "null"}) {
      if (text.startsWith(literal, offset)) {
        offset += literal.length();
        return literal.equals("null") ? null : Boolean.valueOf(literal);
      }
    }
    int start = offset;
    while (offset < text.length() && text.charAt(offset) >= '0' && text.charAt(offset) <= '9') offset++;
    if (start == offset || (offset - start > 1 && text.charAt(start) == '0')) throw error();
    long number = Long.parseLong(text.substring(start, offset));
    if (number > 9007199254740991L) throw error(); return number;
  }
  private static String quote(String text) {
    return "\"" + text.replace("\\", "\\\\").replace("\"", "\\\"") + "\"";
  }
  private static String encode(Object value) {
    if (value == null) return "null";
    if (value instanceof String) return quote((String) value);
    StringBuilder result = new StringBuilder();
    if (value instanceof Map) {
      result.append('{'); boolean first = true;
      for (Map.Entry<?, ?> entry : ((Map<?, ?>) value).entrySet()) {
        if (!first) result.append(','); first = false;
        result.append(quote((String) entry.getKey())).append(':').append(encode(entry.getValue()));
      }
      return result.append('}').toString();
    }
    if (value instanceof List) {
      result.append('['); boolean first = true;
      for (Object item : (List<?>) value) {
        if (!first) result.append(','); first = false; result.append(encode(item));
      }
      return result.append(']').toString();
    }
    return value.toString();
  }
  public static void main(String[] args) throws Exception {
    if (args.length != 1) throw new IllegalArgumentException("usage: Canonical document.json");
    byte[] bytes = Files.readAllBytes(Paths.get(args[0]));
    Canonical parser = new Canonical(new String(bytes, StandardCharsets.UTF_8));
    Object value = parser.value(); parser.space();
    if (parser.offset != parser.text.length()) throw parser.error();
    System.out.write((encode(value) + "\n").getBytes(StandardCharsets.US_ASCII));
  }
}
